staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go

```diff
@@ -17,688 +17,16 @@ limitations under the License.
-	}
-	if t.Size() != oMap.Size() {
-		return types.False
-	}
-	for key, value := range t.value {
-		if propSchema, ok := t.propSchema(key); ok {
-			ov, found := oMap.Find(types.String(key))
-			if !found {
-				return types.False
-			}
-			v := UnstructuredToVal(value, propSchema)
-			vEq := v.Equal(ov)
-			if vEq != types.True {
-				return vEq // either false or error
-			}
-		} else {
-			// Must be an object with properties.
-			// Since we've encountered an unknown field, fallback to unstructured equality checking.
-			ouMap, ok := other.(*unstructuredMap)
-			if !ok {
-				// The compiler ensures equality is against the same type of object, so this should be unreachable
-				return types.MaybeNoSuchOverloadErr(other)
-			}
-			if oValue, ok := ouMap.value[key]; ok {
-				if !equality.Semantic.DeepEqual(value, oValue) {
-					return types.False
-				}
-			}
-		}
-	}
-	return types.True
-}
-
-func (t *unstructuredMap) Type() ref.Type {
-	return types.MapType
-}
-
-func (t *unstructuredMap) Value() interface{} {
-	return t.value
-}
-
-func (t *unstructuredMap) Contains(key ref.Val) ref.Val {
-	v, found := t.Find(key)
-	if v != nil && types.IsUnknownOrError(v) {
-		return v
-	}
-
-	return types.Bool(found)
-}
-
-func (t *unstructuredMap) Get(key ref.Val) ref.Val {
-	v, found := t.Find(key)
-	if found {
-		return v
-	}
-	return types.ValOrErr(key, "no such key: %v", key)
-}
-
-func (t *unstructuredMap) Iterator() traits.Iterator {
-	isObject := t.schema.Properties != nil
-	keys := make([]ref.Val, len(t.value))
-	i := 0
-	for k := range t.value {
-		if _, ok := t.propSchema(k); ok {
-			mapKey := k
-			if isObject {
-				if escaped, ok := cel.Escape(k); ok {
-					mapKey = escaped
-				}
-			}
-			keys[i] = types.String(mapKey)
-			i++
-		}
-	}
-	return &mapIterator{unstructuredMap: t, keys: keys}
-}
-
-type mapIterator struct {
-	*unstructuredMap
-	keys []ref.Val
-	idx  int
-}
-
-func (it *mapIterator) HasNext() ref.Val {
-	return types.Bool(it.idx < len(it.keys))
-}
-
-func (it *mapIterator) Next() ref.Val {
-	key := it.keys[it.idx]
-	it.idx++
-	return key
-}
-
-func (t *unstructuredMap) Size() ref.Val {
-	return types.Int(len(t.value))
-}
-
-func (t *unstructuredMap) Find(key ref.Val) (ref.Val, bool) {
-	isObject := t.schema.Properties != nil
-	keyStr, ok := key.(types.String)
-	if !ok {
-		return types.MaybeNoSuchOverloadErr(key), true
-	}
-	k := keyStr.Value().(string)
-	if isObject {
-		k, ok = cel.Unescape(k)
-		if !ok {
-			return nil, false
-		}
-	}
-	if v, ok := t.value[k]; ok {
-		// If this is an object with properties, not an object with additionalProperties,
-		// then null valued nullable fields are treated the same as absent optional fields.
-		if isObject && v == nil {
-			return nil, false
-		}
-		if propSchema, ok := t.propSchema(k); ok {
-			return UnstructuredToVal(v, propSchema), true
-		}
-	}
-
-	return nil, false
+	return celopenapi.UnstructuredToVal(unstructured, schema.ToKubeOpenAPI())
```

Comment:
same question about whether ToKubeOpenAPI() is called once per validator instance or once per validation

Reply:
The conversion is per validator, which can be expensive especially for a deeply nested schema. I've changed them to use a adaptor to avoid the conversion.